File: packages/typescript/src/resources/contact-groups.ts
/**
 * The `contactGroups` resource: named sets of contacts and group sends.
 * @module
 */
import { toWire } from '../models.js';
import type {
  Batch,
  ContactGroup,
  ContactGroupSendParams,
  CreateContactGroupParams,
  ListParams,
  Page,
  RequestOptions,
  UpdateContactGroupParams,
} from '../models.js';
import { Resource, seg, wireQuery } from './base.js';

/** Accessed as `client.contactGroups`. */
export class ContactGroups extends Resource {
  list(params: ListParams = {}, options?: RequestOptions): Promise<Page<ContactGroup>> {
    return this.page({ method: 'GET', path: '/v1/contact-groups', query: wireQuery(params) }, options);
  }

  create(params: CreateContactGroupParams, options?: RequestOptions): Promise<ContactGroup> {
    return this.one({ method: 'POST', path: '/v1/contact-groups', json: toWire(params), idempotent: true }, options);
  }

  get(id: string, options?: RequestOptions): Promise<ContactGroup> {
    return this.one({ method: 'GET', path: `/v1/contact-groups/${seg(id)}` }, options);
  }

  update(id: string, params: UpdateContactGroupParams, options?: RequestOptions): Promise<ContactGroup> {
    return this.one({ method: 'PATCH', path: `/v1/contact-groups/${seg(id)}`, json: toWire(params) }, options);
  }

  delete(id: string, options?: RequestOptions): Promise<void> {
    return this.none({ method: 'DELETE', path: `/v1/contact-groups/${seg(id)}` }, options);
  }

  /** Send `text` or a template to every contact in the group. Returns a running Batch. */
  send(id: string, params: ContactGroupSendParams, options?: RequestOptions): Promise<Batch> {
    return this.one(
      { method: 'POST', path: `/v1/contact-groups/${seg(id)}/send`, json: toWire(params), idempotent: true },
      options,
    );
  }
}

File: packages/rust/src/resources/contact_groups.rs
//! The `contact_groups` resource.

use super::{idem, list_query, seg};
use crate::error::Result;
use crate::http::{HttpClient, Idem};
use crate::models::{
    Batch, ContactGroup, CreateContactGroup, ListParams, Page, RequestOptions, SendToGroup,
    UpdateContactGroup,
};

/// Accessed as `client.contact_groups()`.
#[derive(Debug, Clone)]
pub struct ContactGroups {
    http: HttpClient,
}

impl ContactGroups {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List groups.
    pub async fn list(&self, params: ListParams) -> Result<Page<ContactGroup>> {
        self.http
            .get(&format!("/v1/contact-groups{}", list_query(&params)))
            .await
    }

    /// Create a group.
    pub async fn create(&self, params: &CreateContactGroup) -> Result<ContactGroup> {
        self.create_with(params, &RequestOptions::default()).await
    }

    /// [`ContactGroups::create`] with an explicit Idempotency-Key.
    pub async fn create_with(
        &self,
        params: &CreateContactGroup,
        opts: &RequestOptions,
    ) -> Result<ContactGroup> {
        self.http
            .send("POST", "/v1/contact-groups", params, idem(opts))
            .await
    }

    /// Fetch a group.
    pub async fn get(&self, id: &str) -> Result<ContactGroup> {
        self.http
            .get(&format!("/v1/contact-groups/{}", seg(id, "id")?))
            .await
    }

    /// Update a group.
    pub async fn update(&self, id: &str, params: &UpdateContactGroup) -> Result<ContactGroup> {
        let path = format!("/v1/contact-groups/{}", seg(id, "id")?);
        self.http.send("PATCH", &path, params, Idem::None).await
    }

    /// Delete a group.
    pub async fn delete(&self, id: &str) -> Result<()> {
        let path = format!("/v1/contact-groups/{}", seg(id, "id")?);
        self.http.empty("DELETE", &path, Idem::None).await
    }

    /// Send to every member. Returns a batch that is already `running`.
    pub async fn send(&self, id: &str, params: &SendToGroup) -> Result<Batch> {
        self.send_with(id, params, &RequestOptions::default()).await
    }

    /// [`ContactGroups::send`] with an explicit Idempotency-Key.
    pub async fn send_with(
        &self,
        id: &str,
        params: &SendToGroup,
        opts: &RequestOptions,
    ) -> Result<Batch> {
        let path = format!("/v1/contact-groups/{}/send", seg(id, "id")?);
        self.http.send("POST", &path, params, idem(opts)).await
    }
}
